rust/lance-encoding/src/encodings/physical/binary.rs

```diff
@@ -0,0 +1,199 @@
+// SPDX-License-Identifier: Apache-2.0
+// SPDX-FileCopyrightText: Copyright The Lance Authors
+
+use std::sync::Arc;
+
+use arrow_array::{Array, ArrayRef, BooleanArray, StringArray, UInt32Array, UInt8Array};
+use arrow_buffer::BooleanBuffer;
+use futures::{future::BoxFuture, FutureExt};
+use log::trace;
+// use rand::seq::index;
+
+use crate::{
+    decoder::{PhysicalPageDecoder, PhysicalPageScheduler},
+    encoder::{ArrayEncoder, BufferEncoder, EncodedArray, EncodedArrayBuffer},
+    format::pb,
+    EncodingsIo,
+};
+
+use lance_core::Result;
+
+use super::buffers::BitmapBufferEncoder;
+
+#[derive(Debug)]
+pub struct BinaryEncoder {
+    indices_encoder: Box<dyn ArrayEncoder>,
+    bytes_encoder: Box<dyn ArrayEncoder>,
+}
+
+impl BinaryEncoder {
+    pub fn new(
+        indices_encoder: Box<dyn ArrayEncoder>,
+        bytes_encoder: Box<dyn ArrayEncoder>,
+    ) -> Self {
+        Self {
+            indices_encoder,
+            bytes_encoder,
+        }
+    }
+}
+
+fn get_indices_from_string_arrays(arrays: &[ArrayRef]) -> Vec<ArrayRef> {
+    let index_arrays: Vec<ArrayRef> = arrays
+        .iter()
+        .filter_map(|arr| {
+            if let Some(string_arr) = arr.as_any().downcast_ref::<StringArray>() {
+                let mut int_arr = Vec::new();
+
+                let mut cum_sz: u32 = 0;
+                for i in 0..string_arr.len() {
+                    let s = string_arr.value(i);
+                    let sz = s.len() as u32;
+                    cum_sz += sz;
+                    int_arr.push(cum_sz);
+                }
+                // Arc::new(UInt32Array::from(int_arr));
+                Some(Arc::new(UInt32Array::from(int_arr)) as ArrayRef)
```

Comment:
`schedule_ranges()`, `update_capacity()`, and `decode_into()` still require u32 values as input. So we may need to change these in order to decode the bytes correctly. I haven't made the change for now since it requires too many reverse casts in other places